gapic/src/main/com/google/gapid/views/LogView.java

```diff
@@ -53,12 +82,107 @@ public LogView(Composite parent) {
   }
 
   private void updateText() {
-    String message = Logging.getLogMessages();
-    text.getDocument().set(message);
-    text.setTopIndex(message.length() - 1);
+    tree.removeAll();
+    for (Log.Message message : Logging.getLogMessages()) {
+      Log.Severity severity = message.getSeverity();
+      TreeItem item = newItem(tree, severity);
+      String[] lines = message.getText().split("(\n)|(\r\n)");
+      item.setText(Column.SEVERITY.index, message.getSeverity().name().substring(0, 1));
+      item.setText(Column.TIME.index, formatTime(message.getTime()));
+      item.setText(Column.PROCESS.index, message.getProcess());
+      item.setText(Column.TEXT.index, lines[0]);
+      item.setText(Column.TAG.index, message.getTag());
+      // Additional lines
+      for (int i = 1; i < lines.length; i++) {
+        TreeItem line = newItem(item, severity);
+        line.setText(Column.TEXT.index, lines[i]);
+      }
+      // Values
+      if (message.getValuesCount() > 0) {
+        TreeItem values = newItem(item, severity);
+        values.setText(2, "Values");
+        for (Log.Value value : message.getValuesList()) {
+          TreeItem entry = newItem(values, severity);
+          String name = value.getName();
+          String val = Pods.unpod(value.getValue()).toString();
+          entry.setText(2, name);
+          entry.setText(Column.TEXT.index, val);
+        }
+      }
+      // Callstack
+      if (message.getCallstackCount() > 0) {
+        TreeItem callstack = newItem(item, severity);
+        callstack.setText(2, "Call Stack");
+        for (Log.SourceLocation location : message.getCallstackList()) {
+          TreeItem line = newItem(callstack, severity);
+          line.setText(Column.TEXT.index, String.format("%s:%d", location.getFile(), location.getLine()));
+        }
+      }
+    }
     dirty.set(false);
   }
 
+  private String formatTime(Timestamp time) {
+    Date date = new Date(time.getSeconds() * 1000);
+    Calendar calendar = Calendar.getInstance();
+    calendar.setTime(date);
+    int hour = calendar.get(Calendar.HOUR);
+    int minute = calendar.get(Calendar.MINUTE);
+    int second = calendar.get(Calendar.SECOND);
+    int millis = time.getNanos() / 1000000;
+    return String.format("%02d:%02d:%02d.%03d", hour, minute, second, millis);
+  }
+
+  private TreeItem newItem(Tree parent, Log.Severity severity) {
+    TreeItem item = new TreeItem(parent, 0);
+    item.setBackground(severityBackgroundColor(severity));
+    item.setForeground(severityForegroundColor(severity));
+    return item;
+  }
+
+  private TreeItem newItem(TreeItem parent, Log.Severity severity) {
+    TreeItem item = new TreeItem(parent, 0);
+    item.setBackground(severityBackgroundColor(severity));
+    item.setForeground(severityForegroundColor(severity));
+    return item;
+  }
+
+  private Color severityBackgroundColor(Log.Severity severity) {
+    switch (severity) {
+      case Verbose:
+        return theme.verboseBackground();
```

Comment:
nit: since the case statements are so simple, I'd prefer them to be on one line:
```
      case Verbose: return theme.verboseBackground();
      case Debug: return theme.debugBackground();
      ...
```